### PR TITLE
Walk directories in filesystem source enumeration

### PR DESCRIPTION
**Pull request by mcastorina**
### Description:
Previously the unit could be a directory, but now it is individual files. This leverages the SourceManager's concurrent units to speed up scanning of large directories.

```
» find /tmp/files -type f | wc -l
  100000

# before
» ./trufflehog filesystem /tmp/files
2024-01-18T12:48:05-08:00       info-0  trufflehog      finished scanning       {"chunks": 100000, "bytes": 588895, "verified_secrets": 0, "unverified_secrets": 0, "scan_duration": "27.942921208s"}
# 27.94s

# after
» ./trufflehog filesystem /tmp/files
2024-01-18T12:47:04-08:00       info-0  trufflehog      finished scanning       {"chunks": 100000, "bytes": 588895, "verified_secrets": 0, "unverified_secrets": 0, "scan_duration": "14.342492916s"}
# 14.34s
```

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/sources/filesystem/filesystem.go [hunk 1]
```diff
@@ -223,6 +223,26 @@ func (s *Source) scanFile(ctx context.Context, path string, chunksChan chan *sou
...
+			continue
+		}
+		if fileInfo.IsDir() {
+			return fs.WalkDir(os.DirFS(path), ".", func(relativePath string, d fs.DirEntry, err error) error {
+				if err != nil || relativePath == "." {
+					return nil
+				}
+				fullPath := filepath.Join(path, relativePath)
+				item := sources.CommonSourceUnit{ID: fullPath}
+				if err := reporter.UnitOk(ctx, item); err != nil {
```
Comment (ahrav) [inline pkg/sources/filesystem/filesystem.go:240]: nit: Can we just `return reporter.UnitOK(ctx, item)` ?

**Review by ahrav — APPROVED** [review]

File: pkg/sources/filesystem/filesystem.go [hunk 2]
```diff
@@ -223,9 +218,32 @@ func (s *Source) scanFile(ctx context.Context, path string, chunksChan chan *sou
...
+		}
+		if !fileInfo.IsDir() {
+			item := sources.CommonSourceUnit{ID: path}
+			if err := reporter.UnitOk(ctx, item); err != nil {
+				return err
+			}
+			continue
+		}
+		err = fs.WalkDir(os.DirFS(path), ".", func(relativePath string, d fs.DirEntry, err error) error {
+			if err != nil || d.IsDir() {
```
Comment (rosecodym) [inline pkg/sources/filesystem/filesystem.go:236]: Does this risk throwing away any errors that we were previously capturing?

**Review by rosecodym — APPROVED** [review]
LGTM, other than my one question.

It is occurring to me that we might need to all come together sometime soonish to talk about our philosophy of concurrency - I think that different parts of the codebase are treating the topic differently and we might need to do some reconciliation.

File: pkg/sources/filesystem/filesystem.go (hunk 2 above)
Comment (mcastorina) [reply pkg/sources/filesystem/filesystem.go:236]: For reference, the [WalkDirFunc](https://pkg.go.dev/io/fs#WalkDirFunc) describes the `err` argument:

> The err argument reports an error related to path, signaling that WalkDir will not walk into that directory. The function can decide how to handle that error; as described earlier, returning the error will cause WalkDir to stop walking the entire tree.

Perhaps there's an argument that we should record the error in the job report, but as it stands we currently ignore any errors given by `WalkDir`: https://github.com/trufflesecurity/trufflehog/blob/8380e1713ea5d3fc21c2f1256ab703f2af30b0de/pkg/sources/filesystem/filesystem.go#L112-L116

**Comment by dustin-decker** [comment]
Looks like this caused a regression
https://github.com/trufflesecurity/trufflehog/issues/2350

**Pull request merged**